xtext-gradle-plugin/src/main/java/org/xtext/gradle/tasks/XtextExtension.xtend

```diff
@@ -59,19 +62,28 @@ class XtextExtension {
 		return null
 	}
 
-	def void forceXtextVersion(Configuration dependencies, ()=>String xtextVersion) {
+	def void forceXtextVersion(Configuration dependencies, String xtextVersion) {
 		dependencies.resolutionStrategy.eachDependency [
 			if (requested.group == "org.eclipse.xtext" || requested.group == "org.eclipse.xtend")
-				useVersion(xtextVersion.apply)
+				useVersion(xtextVersion)
 		]
+
+		if (project.supportsJvmEcoSystemplugin && VersionNumber.parse(xtextVersion)>=  VersionNumber.parse("2.17.1")) {
+			dependencies.dependencies += project.dependencies.enforcedPlatform('''org.eclipse.xtext:xtext-dev-bom:«xtextVersion»''')
```

Comment:
this is problematic when the bootstrap version for xtend compilation in xtext-core
we now end up with an "old" bom as we override the bom for the project, not the xtend compiler

Reply:
```
> Task :org.eclipse.xtext.ide:compileJava FAILED
/Users/cdietrich/xtext-master/git/xtext-core/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/contentassist/ContentAssistService.java:19: error: cannot find symbol
import org.eclipse.lsp4j.InsertReplaceEdit;
                        ^
  symbol:   class InsertReplaceEdit
  location: package org.eclipse.lsp4j
/Users/cdietrich/xtext-master/git/xtext-core/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/contentassist/ContentAssistService.java:125: error: cannot find symbol
                completionItem.setTextEdit(Either.<TextEdit,InsertReplaceEdit>forLeft(new TextEdit(new Range(prefixPosition, caretPosition), entry.getProposal())));
```
here an old lsp4j seems to be pulled for java compilation

Reply:
i also still get
```
Caused by: com.google.inject.internal.util.$ComputationException: com.google.inject.internal.util.$ComputationException: com.google.inject.internal.util.$ComputationException: com.google.inject.internal.util.$ComputationException: com.google.inject.internal.util.$ComputationException: java.lang.UnsupportedClassVersionError: org/eclipse/core/runtime/OperationCanceledException has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```
so the force seem not to be called.

Reply:
https://github.com/eclipse/xtext-core/tree/cd_xxxxxxx

Reply:
This logic is only applied to the Xtext classpath that the plugin automatically generates. If you set a custom classpath, you have to make sure it is set up correctly yourself. So you still need the BOM in that bootstrap use case.

Reply:
hmmm ok so i cannot remove the bom, ok

Reply:
no, it still does not work....

Reply:
```
Execution failed for task ':org.eclipse.xtext.xtext.wizard:generateXtext'.
> java.lang.ClassNotFoundException: org.eclipse.xtext.java.JavaSourceLanguageSetup
```

Reply:
maybe there is a better way todo the bootstrapping. have no idea why it is done as it is done.
i thought xtext.java was a dependendy of xtend.core

Reply:
its actually not, but before the gradle plugin seems to have pulled it.

Reply:
if i add it manually this works. the lsp4j problem still is there

Reply:
Okay I'll have to refine the classpath setup so we always get xtext.java & friends even when someone sets a custom classpath. That might actually fix the other issue as well so you wouldn't strictly need the BOM anymore.

Reply:
what i dont understand is why we in xtext-core configure the resplution strategy AND dependencies for xtendTooling
if i understand correct the dependencies should not be needed.
but if i remove the dependencies section i get
```
Execution failed for task ':org.eclipse.xtext.xtext.generator:generateXtext'.
> No org.xtext.gradle.protocol.IncrementalXtextBuilderFactory found on the classpath
```

Reply:
The dependencies are needed, because you are completely overwriting the xtextClasspath, which means none of the default logic runs (no platform, no xtext.java and other default dependencies). If you do this (which is only needed for the bootstrap use case), you need to take care of all those things yourself.

I have changed the logic so that the Builder is always there (since it's part of the plugin), but everything else needs to be supplied manually in this case, including xtext.java etc. Your branch now builds correctly with my latest change.

Reply:
hmmm i still have problems with lsp4j. how do you build the plugin? i still dont see a publish config checked in.

Reply:
ahh found it in updated readme.

Reply:
`./gradlew pTML -PreleaseVersion=<version>`

Reply:
hmm still have the lsp4j problems when java compile xtext.ide

Reply:
`./gradlew clean build -PuseJenkinsSnapshots=true -PcompileXtend=true`

Reply:
```
> Task :org.eclipse.xtext.ide:compileJava FAILED
/Users/cdietrich/xtext-master/git/xtext-core/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/contentassist/ContentAssistService.java:124: error: incompatible types: no instance(s) of type variable(s) L,R exist so that org.eclipse.lsp4j.jsonrpc.messages.Either<L,R> conforms to org.eclipse.lsp4j.TextEdit
                completionItem.setTextEdit(Either.forLeft(new TextEdit(new Range(prefixPosition, caretPosition), entry.getProposal())));
```

Reply:
```
2021-08-28T18:28:55.793+0200 [DEBUG] [org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler] Compiler arguments: -source 1.8 -target 1.8 -d /Users/cdietrich/xtext-master/git/xtext-core/org.eclipse.xtext.ide/build/classes/java/main -encoding ISO-8859-1 -h /Users/cdietrich/xtext-master/git/xtext-core/org.eclipse.xtext.ide/build/generated/sources/headers/java/main -g -sourcepath  -proc:none -s /Users/cdietrich/xtext-master/git/xtext-core/org.eclipse.xtext.ide/build/generated/sources/annotationProcessor/java/main -XDuseUnsharedTable=true -classpath /Users/cdietrich/xtext-master/git/xtext-core/org.eclipse.xtext/build/classes/java/main:/Users/cdietrich/xtext-master/git/xtext-core/org.eclipse.xtext.util/build/classes/java/main:/Users/cdietrich/.gradle/caches/modules-2/files-2.1/org.eclipse.emf/org.eclipse.emf.ecore.change/2.14.0/919d756fed4322b1dac8ea37e54f186ee8a6d6bd/org.eclipse.emf.ecore.change-2.14.0.jar:/Users/cdietrich/.gradle/caches/modules-2/files-2.1/org.eclipse.lsp4j/org.eclipse.lsp4j/0.10.0/271665493a1e35fb4c2d1dbabf19363c8ad27819/org.eclipse.lsp4j-0.10.0.jar:/Users/cdietrich/.gradle/caches/modules-2/files-2.1/org.eclipse.lsp4j/
```
it seems to pull lsp4j 0.10.0 from the milestone and not 0.12 from snapshots
https://github.com/eclipse/xtext-lib/blob/5d42ac455308f145586d090d6b720bdfe7e4b766/org.eclipse.xtext.dev-bom/build.gradle#L54
need to check if i have some bs lying around in my repos

Reply:
ok it looks like some old dev bom was lying around in my mavenLocal